Added an faq for xcodebuild for devices openssl signing error (#820)
appcircleio/appcircle-docusaurus: docs/workflows/ios-specific-workflow-steps/xcodebuild-for-devices.md

end
```

Now, the `run_command_simple()` function will execute your customized `xcodebuild` command.

> ### How can I resolve OpenSSL signing error? (Only Self-Hosted Users)

Reviewer: I think cloud users might face this issue if they run some custom scripts before the Xcode build step.

Although **LibreSSL** and **OpenSSL** are alternatives to each other, there are differences between them. **LibreSSL** comes by default with macOS machines and is managed by **Apple**. For this reason, since Appcircle does not have direct access to self-hosted environments, some user-side work on runners can replace **LibreSSL** with **OpenSSL** or update their versions. If **OpenSSL** is used instead of **LibreSSL** for any reason in your self-hosted environments, you will get an **error** like below.

```
`parse_certificate': Error outputting keys and certificates (RuntimeError)
C05EDAE401000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
Could not find certificate from <stdin>
Error: Error outputting keys and certificates
C05EDAE401000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
```

> The reason for this **error** is that the **encryption algorithm** in the new versions of **OpenSSL** has been changed. In **OpenSSL** versions **3 and above**, the algorithm named **RC2** is marked as **legacy**. When you encounter this error, you need to change the **OpenSSL** tool on the runners receiving the error to **LibreSSL**. 

Reviewer: RC2 is just an example. There are other deprecated algorithms / ciphers. Users can face other error messages for other algorithms like SHA1. That depends on the user signing certificate.

end
```

Now, the `run_command_simple()` function will execute your customized `xcodebuild` command.

> ### How can I resolve the OpenSSL signing error?

Reviewer 2: Since openSSL is not mentioned in this error message, the user will not search the error for **OpenSSL**. So can we change it as follows?
```suggestion
### How can I solve the "Error outputting keys and certificates"?
```

> All tools running in Appcircle cloud environments are controlled by the Appcircle development teams on runners and updated when necessary. One of the tools used on runners is **LibreSSL**. In Appcircle Cloud environments, the **LibreSSL** 3.3.6 version on macOS Sonoma and the **LibreSSL** 2.8.3 version on macOS Monterey are used. For more information, please visit our [**Build Infrastructure**](/infrastructure/ios-build-infrastructure#ios-build-agent-stacks) documentations.

Reviewer 2: IMO this info should be in :
:::info Cloud Customers

Although **LibreSSL** and **OpenSSL** are alternatives to each other, there are differences between them. **LibreSSL** comes by default with macOS machines and is managed by **Apple**. For this reason, since Appcircle does not have direct access to self-hosted environments, some user-side work on runners can replace **LibreSSL** with **OpenSSL** or update their versions. If **OpenSSL** is used instead of **LibreSSL** for any reason in your self-hosted environments, you will get an **error** like the one below.

> :::info Cloud Customers

Reviewer 2: IMO, this information should be under the error message. Because the error message should be at the top of the FAQ so users can see whether it fitst their error message or not.

Cloud users may also encounter such errors. If you are working in the cloud environment and the [**Custom Scripts**](/workflows/common-workflow-steps/custom-script) you use can **change** or **update** the tools in our environments. If cloud users encounter such **signing errors**, it is recommended to check the **Custom Scripts** used.

:::

```
`parse_certificate': Error outputting keys and certificates (RuntimeError)
C05EDAE401000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
Could not find certificate from <stdin>
Error: Error outputting keys and certificates
C05EDAE401000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
```

> The reason for this **error** is that the **encryption algorithm** in the new versions of **OpenSSL** has been changed. In **OpenSSL** versions **3 and above**, the algorithm named **RC2** is marked as **legacy**. When you encounter this error, you need to change the **OpenSSL** tool on the runners receiving the error to **LibreSSL**. 

Reviewer 2: Can you add a white line after the code block:
```suggestion

The reason for this **error** is that the **encryption algorithm** in the new versions of **OpenSSL** has been changed. In **OpenSSL** versions **3 and above**, the algorithm named **RC2** is marked as **legacy**. When you encounter this error, you need to change the **OpenSSL** tool on the runners receiving the error to **LibreSSL**. 
```

The **RC2 algorithm** is just one example. There are other algorithms and ciphers that **OpenSSL** has deprecated. Users may encounter other errors with certificates containing other algorithms, such as **SHA1**. This depends on the encryption algorithm of the certificate the user is using.

> 

Reviewer 2: Not important one but IMO there is no need two whiteline:
```suggestion
```

> For more information about **legacy algorithms**, please visit the [**OpenSSL**](https://docs.openssl.org/3.0/man7/OSSL_PROVIDER-legacy/) documentations.

Reviewer 2: ```suggestion
For more information about **legacy algorithms**, please visit the [**OpenSSL**](https://docs.openssl.org/3.0/man7/OSSL_PROVIDER-legacy/) documentation.
```

Reviewer 2: @boztopuz , This feedback has not been addresse. Could you please check?

Author: @ftmzhrsckldr I already changed the feedback.

Reviewer 2: Oh sorry there is same problem on another sentence. I will add the suggestion in there 👍🏻

All tools running in Appcircle cloud environments are controlled by the Appcircle development teams on runners and updated when necessary. One of the tools used on runners is **LibreSSL**. In Appcircle Cloud environments, the **LibreSSL** 3.3.6 version on macOS Sonoma and the **LibreSSL** 2.8.3 version on macOS Monterey are used. For more information, please visit our [**Build Infrastructure**](/infrastructure/ios-build-infrastructure#ios-build-agent-stacks) documentations.

Cloud users may also encounter such errors. If you are working in the cloud environment and the [**Custom Scripts**](/workflows/common-workflow-steps/custom-script) you use can **change** or **update** the tools in our environments. If cloud users encounter such **signing errors**, it is recommended to check the **Custom Scripts** used.

:::

Although **LibreSSL** and **OpenSSL** are alternatives to each other, there are differences between them. **LibreSSL** comes by default with macOS machines and is managed by **Apple**. For this reason, since Appcircle does not have direct access to self-hosted environments, some user-side work on runners can replace **LibreSSL** with **OpenSSL** or update their versions.

> The reason for this **error** is that the **encryption algorithm** in the new versions of **OpenSSL** has been changed. In **OpenSSL** versions **3 and above**, the algorithm named **RC2** is marked as **legacy**. When you encounter this error, you need to change the **OpenSSL** tool on the runners receiving the error to **LibreSSL**. 

Reviewer 2: @boztopuz  If this warning is specific to self-hosted environments, could you please clarify that?

Author: No this is general explanation. It is not self-hosted specific.

Reviewer 2: > you need to change the **OpenSSL** tool on the runners

It is written that they need to make changes in the runner, did you mean Custom Script? If so, wouldn't it be more clear to use "you should change this package in the build workflow"? Because cloud users cannot access the runners 👀

Author: To self-hosted customers, they need to change default runner settings. Even they change the tools from CS this is not a default one and may cause other problems. So they should change the default settings of runners in their env. 

This part of the document contains an annotation proving the cause of the problem. You can think of it as a more detailed explanation. While CS can be used for cloud customers and I already explain before, for self-hosted customers the default settings of the runners need to be changed.

> The reason for this **error** is that the **encryption algorithm** in the new versions of **OpenSSL** has been changed. In **OpenSSL** versions **3 and above**, the algorithm named **RC2** is marked as **legacy**. When you encounter this error, you need to change the **OpenSSL** tool on the runners receiving the error to **LibreSSL**. 

Reviewer 2: > you need to change the **OpenSSL** tool on the runners

@boztopuz If there is any documentation explaining how to do this, can we include the link? (Also, should we refer to it as a tool or a package here?)

Author: I already gave a basic script to do this.

Now, the `run_command_simple()` function will execute your customized `xcodebuild` command.

> ### How can I resolve the `Algorithm (RC2-40-CBC : 0)` signing error?

Author: I actually change the title after your feedback. I did not realize that, you are right the exact error message is outputting keys an certificates. 

I used it in the title because it is a very specific error code.

Error: Error outputting keys and certificates
C05EDAE401000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
```

:::info Cloud Customers

All tools running in Appcircle cloud environments are controlled by the Appcircle development teams on runners and updated when necessary. One of the tools used on runners is **LibreSSL**. In Appcircle Cloud environments, the **LibreSSL** 3.3.6 version on macOS Sonoma and the **LibreSSL** 2.8.3 version on macOS Monterey are used. For more information, please visit our [**Build Infrastructure**](/infrastructure/ios-build-infrastructure#ios-build-agent-stacks) documentations.

> Cloud users may also encounter such errors. If you are working in the cloud environment and the [**Custom Scripts**](/workflows/common-workflow-steps/custom-script) you use can **change** or **update** the tools in our environments. If cloud users encounter such **signing errors**, it is recommended to check the **Custom Scripts** used.

Author: I added a basic script to do this

:::info Cloud Customers

> All packages running in Appcircle cloud environments are controlled by the Appcircle development teams on runners and updated when necessary. One of the packages used on runners is **LibreSSL**. In Appcircle Cloud environments, the **LibreSSL** 3.3.6 version on macOS Sonoma and the **LibreSSL** 2.8.3 version on macOS Monterey are used. For more information, please visit our [**Build Infrastructure**](/infrastructure/ios-build-infrastructure#ios-build-agent-stacks) documentations.

Reviewer 2: ```suggestion
All packages running in Appcircle cloud environments are controlled by the Appcircle development teams on runners and updated when necessary. One of the packages used on runners is **LibreSSL**. In Appcircle Cloud environments, the **LibreSSL** 3.3.6 version on macOS Sonoma and the **LibreSSL** 2.8.3 version on macOS Monterey are used. For more information, please visit our [**Build Infrastructure**](/infrastructure/ios-build-infrastructure#ios-build-agent-stacks) documentation.
```
